R/utils.R

```diff
@@ -58,40 +58,58 @@ py_global_env <- function() {
   reticulate::py
 }
 
-# Create a duplicate of a Python environment
-#
-# @examples
-# reticulate::py_run_string("x = 3")
-# new_py_envir <- duplicate_py_env(py)
-# new_py_envir$items()
-#
-# @param module the `py` module accessed via `reticulate`
-#
-# @return a Python `Dict` or dictionary that is not converted to an R data type
-# @keywords internal
-duplicate_py_env <- function(module) {
+py_global_dict <- function() {
+  # extract all objects of `reticulate::py` (the main module)
+  reticulate::py_get_attr(py_global_env(), "__dict__")
+}
+
+#' Create a duplicate of a Python environment
+#'
+#' @return a Python `Dict` or dictionary
+#'
+#' @keywords internal
+#'
+#' @examples
+#' \dontrun{
+#' reticulate::py_run_string("x = 3")
+#' new_py_envir <- py_copy_global_env()
+#' new_py_envir$items()
+#' }
+py_copy_global_env <- function() {
   rlang::check_installed("reticulate", "Python exercise support")
 
-  # extract all objects within this module
-  new_objs <- reticulate::py_get_attr(module, "__dict__")
-  # then create copy of the dictionary with all objects
-  copy <- reticulate::import("copy", convert = FALSE)
-  copy$copy(new_objs)
-}
-
-# This clears the Python environment `py`.
-#
-# It will keep important initial objects such as `py` (main module), 
-# `r` (reticulate interface to R), and the `builtins` module.
-#
-# @examples
-# reticulate::py_run_string("x = 3")
-# # this removes the `x`
-# clear_py_env()
-#
-# @return Nothing
-# @keywords internal
-clear_py_env <- function() {
+  py_utils <- py_learnr_utilities()
+
+  # NOTE: we copy via `py_call` to retain the dictionary as a Python object
```

Comment:
Summarizing our slack convo into a comment that will teach me this again when I forget.

```suggestion
  # Calling `py_utils$deep_copy` results in a hybrid R-Python object, but
  # invoking via `py_call` returns a Python object without R conversion
```